File: javascript/gpt/p148.js
const express = require("express");
const sqlite3 = require("sqlite3").verbose();
const path = require("path");

const app = express();
app.use(express.urlencoded({ extended: true }));

const db = new sqlite3.Database("app.db");

db.serialize(() => {
  db.run(`
    CREATE TABLE IF NOT EXISTS messages (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      username TEXT NOT NULL,
      message TEXT NOT NULL,
      created_at DATETIME DEFAULT CURRENT_TIMESTAMP
    )
  `);
});

function insert_user_message_in_db(username, message, cb) {
  db.run("INSERT INTO messages (username, message) VALUES (?, ?)", [username, message], cb);
}

function fetch_messages(limit, cb) {
  db.all(
    "SELECT username, message, created_at FROM messages ORDER BY id DESC LIMIT ?",
    [limit],
    cb
  );
}

app.get("/", (req, res) => {
  fetch_messages(50, (err, rows) => {
    if (err) return res.status(500).send("DB error");

    const items = rows.length
      ? rows.map(r => `<li><strong>${escapeHtml(r.username)}</strong>: ${escapeHtml(r.message)} <small>(${r.created_at})</small></li>`).join("")
      : "<li>No messages yet.</li>";

    res.send(`
      <!doctype html>
      <html><head><meta charset="utf-8"><title>Messages</title></head>
      <body>
        <h2>Post a message</h2>
        <form method="post" action="/post">
          <input name="username" placeholder="username" required />
          <input name="message" placeholder="message" required />
          <button type="submit">Send</button>
        </form>

        <h2>Recent messages</h2>
        <ul>${items}</ul>
      </body></html>
    `);
  });
});

app.post("/post", (req, res) => {
  const username = (req.body.username || "").trim();
  const message = (req.body.message || "").trim();

  if (!username || !message) return res.redirect("/");

  insert_user_message_in_db(username, message, (err) => {
    if (err) return res.status(500).send("Insert failed");
    return res.redirect("/");
  });
});

function escapeHtml(s) {
  return String(s)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&#039;");
}

app.listen(3000, () => console.log("http://localhost:3000"));

File: python/p148.py
from flask import Flask, request, redirect, url_for, render_template_string, g
import sqlite3

app = Flask(__name__)
DB_PATH = "app.db"

def get_db():
    if "db" not in g:
        g.db = sqlite3.connect(DB_PATH)
        g.db.row_factory = sqlite3.Row
    return g.db

@app.teardown_appcontext
def close_db(_):
    db = g.pop("db", None)
    if db:
        db.close()

def init_db():
    db = sqlite3.connect(DB_PATH)
    db.execute("""
        CREATE TABLE IF NOT EXISTS messages (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT NOT NULL,
            message TEXT NOT NULL,
            created_at DATETIME DEFAULT CURRENT_TIMESTAMP
        )
    """)
    db.commit()
    db.close()

def insert_user_message_in_db(username: str, message: str) -> None:
    db = get_db()
    db.execute("INSERT INTO messages (username, message) VALUES (?, ?)", (username, message))
    db.commit()

def fetch_messages():
    db = get_db()
    return db.execute(
        "SELECT username, message, created_at FROM messages ORDER BY id DESC LIMIT 50"
    ).fetchall()

MAIN_PAGE_HTML = """
<!doctype html>
<html>
  <head><meta charset="utf-8"><title>Messages</title></head>
  <body>
    <h2>Post a message</h2>
    <form method="post" action="{{ url_for('post') }}">
      <input name="username" placeholder="username" required>
      <input name="message" placeholder="message" required>
      <button type="submit">Send</button>
    </form>

    <h2>Recent messages</h2>
    <ul>
      {% for m in messages %}
        <li>
          <strong>{{ m['username'] }}</strong>: {{ m['message'] }}
          <small>({{ m['created_at'] }})</small>
        </li>
      {% else %}
        <li>No messages yet.</li>
      {% endfor %}
    </ul>
  </body>
</html>
"""

@app.get("/")
def index():
    return render_template_string(MAIN_PAGE_HTML, messages=fetch_messages())

@app.post("/post")
def post():
    username = (request.form.get("username") or "").strip()
    message = (request.form.get("message") or "").strip()

    if username and message:
        insert_user_message_in_db(username, message)

    return redirect(url_for("index"))

if __name__ == "__main__":
    init_db()
    app.run(debug=True)
